File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) 
    {
        int n=nums.size();
        vector<int> prifix(n);
        vector<int> sufix(n);  
        
        prifix[0]=nums[0];
        sufix[n-1]=nums[n-1];
        for(int i=1;i<n;i++)
        {
            prifix[i]=max(prifix[i-1],nums[i]);
        } 
        for(int i=n-2;i>=0;i--)
        {
            sufix[i]=min(sufix[i+1],nums[i]);
        }
        int x;
        for(int i=0;i<n;i++)
        {
            x=prifix[i]-sufix[i];
            if(x<=k)
            {
                return i;
            }
        }

       return -1;
    }
};

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)

        p = [0] * n
        s = [0] * n

    
        p[0] = nums[0]

        for i in range(1, n):
            p[i] = max(nums[i], p[i-1])

    
        s[n-1] = nums[n-1]

        for i in range(n-2, -1, -1):
            s[i] = min(nums[i], s[i+1])

        
        for i in range(n):
            x = p[i] - s[i]

            if x <= k:
                return i

        return -1